### ISSUE TITLE
join_authorised_via_users_server is not protected from redactions

### ISSUE THREAD

**Issue by tulir** [matrix-org/matrix-doc#3373]
Room v8 added `join_authorised_via_users_server` to `m.room.member` events to help with authorizing restricted room join events. However, the field isn't protected from redactions, which means the join event will suddenly become invalid (i.e. not pass event auth rules) if it's redacted. At least with synapse, existing servers will still see the member in the room, but any new servers will reject the event and any messages that the user sends.

**Comment by tulir** [comment]
It appears that sending a new join event might partially "fix" it: servers that joined after a new member event was sent (which is not redacted, but doesn't have a `join_authorised_via_users_server` field) seem to reject the first one, but accept the second one (maybe because it's a join -> join transition).

note: I haven't looked into what actually happened inside synapse, so "reject" means "user can't see event"

**Comment by clokep** [comment]
> It appears that sending a new join event might partially "fix" it: servers that joined after a new member event was sent (which is not redacted, but doesn't have a `join_authorised_via_users_server` field) seem to reject the first one, but accept the second one (maybe because it's a join -> join transition).

Only an initial join needs that field, subsequent joins do not need to include it (since `join` -> `join` is an allowed transition, always).

**Comment by turt2live** [comment]
This feels like a small enough oversight where we can patch it into the v8 spec when that gets written. Some rooms might be affected, but considering it's not even in the spec yet I'm not too worried about them (the fix also seems to be to leave and get rejoined).

**Comment by uhoreg** [comment]
I don't think we can just patch it in because event IDs are based on the redaction algorithm, so if we change the redaction algorithm, servers will disagree about the event IDs for those events.

**Comment by turt2live** [comment]
oh, good point.

**Comment by clokep** [comment]
I did some testing with this:

1. Create a public space and a v8 room on server A as alice.
2. Update the join rules of the room to restricted to members of the space.
2. Join the space as bob.
3. Join the room as bob (via being a member of the space).
4. Redact bob's join event.
5. Alice and bob should still be able to send messages, etc.
6. Join the space as charlie on server B.
7. Join the room as charlie.
8. Note that charlie only sees alice and charlie in the room and rejects all new messages from bob.
9. Alice and bob see three members and all messages sent.

So this effectively split brains the room. Going through the above with a patch to avoid redacting the `join_authorised_via_users_server` field everything works as expected.

Unfortunately I think the only recourse here is to abandon the v8 room version and fix it in another one, unless someone has a more clever solution!

**Comment by clokep** [comment]
See #3375 for a fix to this. 😢

**Comment by turt2live** [comment]
For links: v8 was introduced by https://github.com/matrix-org/matrix-doc/pull/3289
